node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -324,8 +324,14 @@ impl<C: Context> ActiveValidator<C> {
             panorama = state.panorama().clone();
         }
         let seq_number = panorama.next_seq_num(state, self.vidx);
-        // TODO: After LNC we won't always need all known endorsements.
-        let endorsed = state.endorsements().collect();
+        let endorsed = panorama
+            .need_endorsements(state)
+            .into_iter()
+            .map(|v| {
+                assert!(state.is_endorsed(v), "Vote must be endorsed");
```

Comment:
Can we really assert that here? We chose `state.panorama()`, which may well contain a `Faulty` entry that we only just learned about and for which we don't have the necessary endorsements yet, I think.

Reply:
Hmm, you're right. It's not guaranteed that we will have them received by the time we create the vote. It is expected that the receiving node will eventually collecte enough endorsements to accept the vote.

Reply:
But we shouldn't add anything to our local protocol state that is still missing endorsements.

Reply:
We already have those votes in our protocol state. Here we are creating a new vote, with votes that we believe should be endorsed eventually. The receiving node will add them to the synchronization queue and wait until all the endorsements this vote depends on are collected and then proceed to vote validation. What I just realized is missing is that `validate_lnc` should check whether we have indeed endorsements for the vote that should be endorsed (according to the vote we're validating).

Reply:
Exactly: I think when creating a new vote, we should cite carefully, so that given the information we already have, we _know_ that all the dependencies exist.

Because imagine we create `A` citing `B` citing `C`, where `C` is by an equivocator, even though we don't have enough endorsements for `B` yet. Then it turns out `B`'s creator is also malicious! Then `B` will never get endorsed and `A` will never satisfy the LNC.